File: sdk/typescript/src/jev.ts
import type { Client, Json, ObjectValue, RecordV1 } from "./index.js";

export type JevAnswer =
  | { type: "noul"; noul: number }
  | { type: "choice"; choice: string; probabilities: Record<string, number>; confidence: number }
  | { type: "score"; score: number; legend: Record<string, string | ObjectValue | Json[]>; probabilities: Record<string, number>; confidence: number };
export interface JevResponse { model: string; answers: Record<string, JevAnswer>; usage?: ObjectValue }
export interface JevRequest { model: string; state: Json; questions: Record<string, { type: "noul" | "choice" | "score"; instructions?: Json; criteria?: Json }> }
const encoder = new TextEncoder();
const prob = (n: unknown): n is number => typeof n === "number" && Number.isFinite(n) && n >= 0 && n <= 1;
const object = (v: unknown): v is Record<string, unknown> => v !== null && typeof v === "object" && !Array.isArray(v);
function validateAnswers(answers: unknown): asserts answers is Record<string, JevAnswer> {
  if (!object(answers) || !Object.keys(answers).length || Object.keys(answers).length > 64) throw new TypeError("Expected 1–64 Jev answers");
  for (const [name, a] of Object.entries(answers)) {
    if (!name || encoder.encode(name).length > 128 || !object(a)) throw new TypeError("Invalid Jev answer");
    if (a.type === "noul") { if (!prob(a.noul)) throw new TypeError("Invalid Noul probability"); continue; }
    if (a.type !== "choice" && a.type !== "score") throw new TypeError("Unknown Jev answer type");
    const p = a.probabilities;
    if (!object(p) || !Object.keys(p).length || Object.keys(p).length > 255 || !prob(a.confidence)
      || Object.entries(p).some(([k, n]) => !k || encoder.encode(k).length > 256 || !prob(n))
      || Math.abs(Object.values(p).reduce<number>((n, v) => n + Number(v), 0) - 1) > 0.001) throw new TypeError("Invalid Jev probability distribution");
    if (a.type === "choice") {
      if (typeof a.choice !== "string" || !Object.hasOwn(p, a.choice)) throw new TypeError("Invalid Jev choice");
    } else {
      const legend = a.legend;
      if (!object(legend) || Object.keys(legend).length < 2 || Object.keys(legend).length > 10
        || Object.keys(legend).length !== Object.keys(p).length
        || Object.entries(legend).some(([k, v]) => !/^\d+$/.test(k) || Number(k) > 4294967295 || (typeof v !== "string" && !object(v) && !Array.isArray(v)) || !Object.hasOwn(p, k))
        || typeof a.score !== "number" || !Number.isFinite(a.score)
        || a.score < Math.min(...Object.keys(legend).map(Number)) || a.score > Math.max(...Object.keys(legend).map(Number))) throw new TypeError("Invalid Jev score/legend");
    }
  }
}

/** Normalize TypeSafe's JSON response. No inference, retries or provider keys here. */
export async function jevDecision(response: JevResponse, options: {
  request: JevRequest; src: string; dst: string; timestampUs: string; mode: "live" | "fixture";
  client?: Client; attachInputs?: boolean; episode?: string;
}): Promise<RecordV1> {
  const { request, src, dst, timestampUs, mode, client, attachInputs = false, episode } = options;
  validateAnswers(response.answers);
  if ([request.model, response.model].some(m => typeof m !== "string" || !m || encoder.encode(m).length > 128)
    || !["live", "fixture"].includes(mode) || !object(request.questions) || request.state === undefined
    || Object.keys(request.questions).length !== Object.keys(response.answers).length
    || Object.entries(response.answers).some(([k, a]) => !Object.hasOwn(request.questions, k) || request.questions[k].type !== a.type)) throw new TypeError("Request/response provenance mismatch");
  for (const id of [src, dst]) if (typeof id !== "string" || !/^\d+$/.test(id) || BigInt(id) > 18446744073709551615n) throw new TypeError("IDs must be u64 decimal strings");
  if (typeof timestampUs !== "string" || !/^-?\d+$/.test(timestampUs) || BigInt(timestampUs) < -9223372036854775808n || BigInt(timestampUs) >= 9223372036854775807n) throw new TypeError("Invalid microsecond timestamp");
  const strictStringify = (v: unknown) => JSON.stringify(v, (_k, value) => {
    if (typeof value === "number" && !Number.isFinite(value)) throw new TypeError("Non-finite JSON value");
    return value;
  });
  const data = encoder.encode(strictStringify(request));
  if (data.length > 1024 * 1024) throw new RangeError("Request exceeds 1 MiB");
  const digest = await crypto.subtle.digest("SHA-256", data);
  const assets: Record<string, string> = {};
  if (attachInputs) {
    if (!client) throw new TypeError("Attachment uploads require a Chronograph client");
    for (const [name, bytes] of [["request", data], ["response", encoder.encode(strictStringify(response))]] as const) {
      assets[name] = await client.uploadAsset(bytes, { version: 1, kind: "opaque", encoding: "json", provenance: { provider: "typesafe", mode } });
    }
  }
  return { src, dst, timestamp_us: timestampUs, ...(episode ? { episode } : {}), assets,
    fields: { provider: "typesafe", model: response.model, requested_model: request.model, mode,
      input_sha256: Array.from(new Uint8Array(digest), b => b.toString(16).padStart(2, "0")).join(""),
      answers: response.answers as unknown as ObjectValue, usage: response.usage ?? {} } };
}

File: sdk/python/chronograph_connectors/jev.py
"""TypeSafe Jev decision binding. No inference or credentials enter the database.

Contract: https://docs.typesafe.ai/api (reviewed 2026-09-20).
The caller runs TypeSafe's SDK and passes a response or JSON dictionary here.
"""
import hashlib
import json
import math


def _prob(value):
    return type(value) in (int, float) and math.isfinite(value) and 0 <= value <= 1


def _json(value):
    if hasattr(value, "model_dump"):
        value = value.model_dump(mode="json")
    return json.loads(json.dumps(value, allow_nan=False, ensure_ascii=False))


def validate_answers(answers):
    if not isinstance(answers, dict) or not 1 <= len(answers) <= 64:
        raise ValueError("Jev requires 1–64 named answers")
    for name, answer in answers.items():
        if not isinstance(name, str) or not 1 <= len(name.encode()) <= 128 or not isinstance(answer, dict):
            raise ValueError("Invalid Jev answer")
        kind = answer.get("type")
        if kind == "noul":
            if not _prob(answer.get("noul")):
                raise ValueError("Noul must be a probability in [0,1]")
            continue
        if kind not in ("choice", "score"):
            raise ValueError("Unsupported Jev answer type")
        probs = answer.get("probabilities")
        if (not isinstance(probs, dict) or not 1 <= len(probs) <= 255
                or any(not isinstance(k, str) or not 1 <= len(k.encode()) <= 256 or not _prob(p) for k, p in probs.items())
                or abs(sum(probs.values()) - 1) > 0.001 or not _prob(answer.get("confidence"))):
            raise ValueError("Jev probabilities must sum to one and confidence must be in [0,1]")
        if kind == "choice":
            if not isinstance(answer.get("choice"), str) or answer["choice"] not in probs:
                raise ValueError("Choice must be one of the probability keys")
        else:
            legend, score = answer.get("legend"), answer.get("score")
            if (not isinstance(legend, dict) or set(legend) != set(probs) or not 2 <= len(legend) <= 10
                    or any(not k.isascii() or not k.isdecimal() or int(k) > 2**32 - 1 or not isinstance(v, (str, dict, list)) for k, v in legend.items())
                    or type(score) not in (int, float) or not math.isfinite(score)
                    or not min(map(int, legend)) <= score <= max(map(int, legend))):
                raise ValueError("Score requires matching numeric legend and a value within its range")


def jev_decision(response, *, request, src, dst, timestamp_us, mode, client=None, attach_inputs=False, episode=None):
    """Normalize an official SDK response; optional request/response assets are explicit.

    mode is mandatory: 'live' for an actual provider response, 'fixture' for samples.
    SHA-256 covers compact UTF-8 JSON with the request's insertion order preserved.
    Raw state and questions are not retained unless attach_inputs=True. This
    fingerprint is provenance, not anonymization of low-entropy input.
    """
    request, response = _json(request), _json(response)
    if not isinstance(request, dict) or not isinstance(response, dict):
        raise ValueError("Request and response must be JSON objects")
    model, requested = response.get("model"), request.get("model")
    if any(not isinstance(v, str) or not 1 <= len(v.encode()) <= 128 for v in (model, requested)):
        raise ValueError("Requested and returned model identifiers are required")
    if mode not in ("live", "fixture"):
        raise ValueError("Set mode explicitly to live or fixture")
    questions, answers = request.get("questions"), response.get("answers")
    validate_answers(answers)
    if (not isinstance(questions, dict) or set(questions) != set(answers)
            or any(not isinstance(questions[k], dict) or questions[k].get("type") != a["type"] for k, a in answers.items())
            or "state" not in request):
        raise ValueError("Every answer must match a supplied question and type")
    data = json.dumps(request, separators=(",", ":"), ensure_ascii=False, allow_nan=False).encode()
    if len(data) > 1024 * 1024:
        raise ValueError("Example request exceeds 1 MiB; preprocess or split your state")
    assets = {}
    if attach_inputs:
        if client is None:
            raise ValueError("A Chronograph client is required for attachment uploads")
        for name, raw in (("request", data), ("response", json.dumps(response, separators=(",", ":"), ensure_ascii=False, allow_nan=False).encode())):
            assets[name] = client.asset(raw, kind="opaque", encoding="json", provenance={"provider": "typesafe", "mode": mode})
    for value, lo, hi in ((src, 0, 2**64 - 1), (dst, 0, 2**64 - 1), (timestamp_us, -2**63, 2**63 - 2)):
        if isinstance(value, bool) or not isinstance(value, (int, str)) or not lo <= int(value) <= hi:
            raise ValueError("Use exact integer IDs and microsecond timestamps")
    return {"src": str(src), "dst": str(dst), "timestamp_us": str(timestamp_us), "episode": episode,
            "assets": assets, "fields": {"provider": "typesafe", "model": model, "requested_model": requested,
            "input_sha256": hashlib.sha256(data).hexdigest(), "answers": answers, "mode": mode,
            "usage": response.get("usage", {})}}
